snowflake-api/src/migration.rs

```diff
@@ -0,0 +1,282 @@
+#![allow(clippy::cast_sign_loss)]
+use arrow::array::{
+    Int32Array, Int64Array, Int8Array, PrimitiveArray, RecordBatch, StringArray, StructArray,
+};
+use arrow::datatypes::ArrowPrimitiveType;
+use async_trait::async_trait;
+use refinery_core::traits::r#async::{AsyncMigrate, AsyncQuery, AsyncTransaction};
+use refinery_core::Migration;
+
+use serde::{de, Deserialize};
+use sqlparser::dialect::SnowflakeDialect;
+
+use tap::Tap;
+use thiserror::Error;
+use time::format_description::well_known::Rfc3339;
+use time::OffsetDateTime;
+
+use crate::{QueryResult, SnowflakeApi, SnowflakeApiError};
+
+const DIALECT: SnowflakeDialect = sqlparser::dialect::SnowflakeDialect {};
+
+/// copied from `refinery_core`
+#[allow(dead_code)]
+#[derive(Debug, Deserialize)]
+enum State {
+    Applied,
+    Unapplied,
+}
+
+/// copied from `refinery_core`
+#[allow(dead_code)]
+#[derive(Debug, Deserialize)]
+enum TypeInner {
+    Versioned,
+    Unversioned,
+}
+
+/// copied from `refinery_core`
+#[allow(dead_code)]
+#[derive(Debug, Deserialize)]
+struct MigrationInner {
+    state: State,
+    name: String,
+    checksum: u64,
+    version: i32,
+    prefix: TypeInner,
+    sql: Option<String>,
+    #[serde(deserialize_with = "deserialize_date")]
+    applied_on: Option<OffsetDateTime>,
+}
+
+fn deserialize_date<'de, D>(deserializer: D) -> Result<Option<OffsetDateTime>, D::Error>
+where
+    D: de::Deserializer<'de>,
+{
+    let s: Option<String> = Option::deserialize(deserializer)?;
+    match s {
+        Some(s) => Ok(Some(
+            OffsetDateTime::parse(&s, &Rfc3339).map_err(de::Error::custom)?,
+        )),
+        None => Ok(None),
+    }
+}
+
+impl MigrationInner {
+    fn applied(
+        version: i32,
+        name: String,
+        applied_on: OffsetDateTime,
+        checksum: u64,
+    ) -> MigrationInner {
+        MigrationInner {
+            state: State::Applied,
+            name,
+            checksum,
+            version,
+            // applied migrations are always versioned
+            prefix: TypeInner::Versioned,
+            sql: None,
+            applied_on: Some(applied_on),
+        }
+    }
+}
+
+impl From<MigrationInner> for Migration {
+    fn from(inner: MigrationInner) -> Self {
+        assert_eq!(
+            std::mem::size_of::<Migration>(),
+            std::mem::size_of::<MigrationInner>()
+        );
+        unsafe { std::mem::transmute(inner) }
+    }
+}
+
+#[async_trait]
+impl AsyncTransaction for SnowflakeApi {
+    type Error = SnowflakeApiError;
+
+    async fn execute(&mut self, queries: &[&str]) -> Result<usize, Self::Error> {
+        self.exec("BEGIN TRANSACTION").await?;
+
+        for query in queries {
+            let parsed_query = sqlparser::parser::Parser::parse_sql(&DIALECT, query)
```

Comment:
from what I can see sqlparser dialect doesn't allow for put/get queries or any session control queries (alter session), it seem to follow public api, which is of limited use for this particular library

Reply:
Ack yeah, you are totally right. Limited snowflake extended dialect coverage means this is worse than a more naiive impl of statement split.